p2p/src/main/kotlin/maru/p2p/discovery/DiscoveryService.kt

```diff
@@ -0,0 +1,187 @@
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+package maru.p2p.discovery
+
+import java.net.InetSocketAddress
+import java.time.Duration
+import java.util.function.Consumer
+import maru.config.P2P
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import org.apache.tuweni.bytes.Bytes
+import org.apache.tuweni.bytes.Bytes32
+import org.apache.tuweni.crypto.SECP256K1.SecretKey
+import org.apache.tuweni.units.bigints.UInt64
+import org.ethereum.beacon.discovery.DiscoverySystem
+import org.ethereum.beacon.discovery.DiscoverySystemBuilder
+import org.ethereum.beacon.discovery.schema.EnrField
+import org.ethereum.beacon.discovery.schema.NodeRecord
+import org.ethereum.beacon.discovery.schema.NodeRecordBuilder
+import org.ethereum.beacon.discovery.schema.NodeRecordFactory
+import org.hyperledger.besu.metrics.noop.NoOpMetricsSystem
+import tech.pegasys.teku.infrastructure.async.Cancellable
+import tech.pegasys.teku.infrastructure.async.MetricTrackingExecutorFactory
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner
+import tech.pegasys.teku.networking.p2p.discovery.discv5.SecretKeyParser
+
+class DiscoveryService(
+  private val privateKeyBytes: ByteArray,
+  private val p2pConfig: P2P,
+) {
+  companion object {
+    val BOOTNODE_REFRESH_DELAY: Duration = Duration.ofMinutes(2)
+  }
+
+  private val log: Logger = LogManager.getLogger(this::javaClass)
+
+  private lateinit var discoverySystem: DiscoverySystem
+
+  private val privateKey = SecretKeyParser.fromLibP2pPrivKey(Bytes.wrap(privateKeyBytes))
+  private val localNodeRecord = localNodeRecord(privateKey = privateKey)
+
+  val delayedExecutor =
+    ScheduledExecutorAsyncRunner.create(
+      "DiscoveryService",
+      1,
+      1,
+      5,
+      MetricTrackingExecutorFactory(NoOpMetricsSystem()),
+    )
+  var bootnodeRefreshTask: Cancellable? = null
+
+  fun start() {
+    val discoveryNetworkBuilder = DiscoverySystemBuilder()
+
+    val bootnodes =
+      p2pConfig.bootnodes
+        .stream()
+        .map { NodeRecordFactory.DEFAULT.fromEnr(it) }
+        .toList()
+
+    discoveryNetworkBuilder.listen(p2pConfig.ipAddress, p2pConfig.discoveryPort.toInt())
+    discoveryNetworkBuilder.secretKey(privateKey)
+    discoveryNetworkBuilder.localNodeRecord(localNodeRecord)
+    discoveryNetworkBuilder.bootnodes(bootnodes)
+
+    discoverySystem = discoveryNetworkBuilder.build()
```

Comment:
Can we build it during `DiscoveryService` construction? This way `discoverySystem` won't be a var, which is good and in case there are configuration errors, we'll fail faster

Reply:
Done